cpp/include/cudf/io/text/detail/trie.hpp

```diff
@@ -0,0 +1,242 @@
+  device_span<trie_node const> _nodes;
+
+  constexpr multistate transition_init(char c)
+  {
+    auto result = multistate();
+
+    result.enqueue(0, 0);
+
+    for (uint8_t curr = 0; curr < _nodes.size() - 1; curr++) {
+      transition_enqueue_all(c, result, curr, curr);
+    }
+    return result;
+  }
+
+  constexpr multistate transition(char c, multistate const& states)
+  {
+    auto result = multistate();
+
+    result.enqueue(0, 0);
+
+    for (uint8_t i = 0; i < states.size(); i++) {
+      transition_enqueue_all(c, result, states.get_head(i), states.get_tail(i));
+    }
+
+    return result;
+  }
+
+  constexpr void transition_enqueue_all(  //
+    char c,
+    multistate& states,
+    uint8_t head,
+    uint8_t curr)
+  {
+    for (uint32_t tail = _nodes[curr].child_begin; tail < _nodes[curr + 1].child_begin; tail++) {
+      if (_nodes[tail].token == c) {  //
+        states.enqueue(head, tail);
+      }
+    }
+  }
+
+  constexpr bool is_match(uint16_t idx) { return static_cast<bool>(get_match_length(idx)); }
+  constexpr uint8_t get_match_length(uint16_t idx) { return _nodes[idx].match_length; }
+
+  template <uint32_t N>
+  constexpr uint8_t get_match_length(multistate const& states)
+  {
+    int8_t val = 0;
+    for (uint8_t i = 0; i < states.size(); i++) {
+      auto match_length = get_match_length(states.get_tail(i));
+      if (match_length > val) { val = match_length; }
+    }
+    return val;
+  }
+};
+
+/**
+ * @brief A flat trie contained in device memory.
+ */
+struct trie {
+ private:
+  cudf::size_type _max_duplicate_tokens;
+  rmm::device_uvector<trie_node> _nodes;
+
+  trie(cudf::size_type max_duplicate_tokens, rmm::device_uvector<trie_node>&& nodes)
+    : _max_duplicate_tokens(max_duplicate_tokens), _nodes(std::move(nodes))
+  {
+  }
+
+  /**
+   * @brief Used to build a hierarchical trie which can then be flattened.
+   */
+  struct trie_builder_node {
+    uint8_t match_length;
+    std::unordered_map<char, std::unique_ptr<trie_builder_node>> children;
+
+    /**
+     * @brief Insert the string in to the trie tree, growing the trie as necessary
+     */
+    void insert(std::string s) { insert(s.c_str(), s.size(), 0); }
+
+   private:
+    trie_builder_node& insert(char const* s, uint16_t size, uint8_t depth)
+    {
+      if (size == 0) {
+        match_length = depth;
+        return *this;
+      }
+
+      if (children[*s] == nullptr) { children[*s] = std::make_unique<trie_builder_node>(); }
+
+      return children[*s]->insert(s + 1, size - 1, depth + 1);
+    }
+  };
+
+ public:
+  /**
+   * @brief Gets the number of nodes contained in this trie.
+   */
+  cudf::size_type size() const { return _nodes.size(); }
+
+  /**
+   * @brief A pessimistic count of duplicate tokens in the trie. Used to determine the maximum
+   * possible stack size required to compute matches of this trie in parallel.
+   */
+  cudf::size_type max_duplicate_tokens() const { return _max_duplicate_tokens; }
```

Comment:
Couldn't it be that the maximum stack size will exceed the number of `_max_duplicate_tokens`?